### PR TITLE
Added #state observable to the Editor class.

### LINKED ISSUE #124: Introduce editor.status (or state)
We should introduce `editor.status` observable property, similar to the CKEditor 4: https://docs.ckeditor.com/ckeditor4/latest/api/CKEDITOR_editor.html#property-status

I propose to have the following possible status values:
 - initializing - during initialization chain.
 - ready - after initializing chain, the editor is ready to work.
 - destroyed - after destroy, editor and all plugins are destroyed.

This is important because there is no way to stop promises when the editor is destroyed and some plugins should check the editor status in `then`.

It can be alternatively called `state`, both `state` and `status` are fine for me. cc @Reinmar

### LINKED ISSUE #2935: Introduce editor.status (or state)
We should introduce `editor.status` observable property, similar to the CKEditor 4: https://docs.ckeditor.com/ckeditor4/latest/api/CKEDITOR_editor.html#property-status

I propose to have the following possible status values:
 - initializing - during initialization chain.
 - ready - after initializing chain, the editor is ready to work.
 - destroyed - after destroy, editor and all plugins are destroyed.

This is important because there is no way to stop promises when the editor is destroyed and some plugins should check the editor status in `then`.

It can be alternatively called `state`, both `state` and `status` are fine for me. cc @Reinmar

### PR DESCRIPTION
**Pull request by oskarwrobel**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Added `#state` observable to the `Editor` class. Closes ckeditor/ckeditor5#2935.

---

~### Requires~
~- https://github.com/ckeditor/ckeditor5-editor-classic/tree/t/ckeditor5-core/124~
~- https://github.com/ckeditor/ckeditor5-editor-balloon/tree/t/ckeditor5-core/124~
~- https://github.com/ckeditor/ckeditor5-editor-inline/tree/t/ckeditor5-core/124~
~- https://github.com/ckeditor/ckeditor5-editor-decoupled/tree/t/ckeditor5-core/124~

### CODE REVIEW

File: src/editor/editor.js [hunk 1]
```diff
@@ -104,6 +104,19 @@ export default class Editor {
 		 */
 		this.t = this.locale.t;
 
+		/**
+		 * Indicates editor initialization status.
+		 *
+		 * The following statuses are available:
```
Comment (Reinmar) [inline src/editor/editor.js]: Blank line after a paragraph

File: src/editor/editor.js [hunk 2]
```diff
@@ -104,6 +104,19 @@ export default class Editor {
...
+		 * * initializing -  during initialization chain
```
Comment (Reinmar) [inline src/editor/editor.js]: A person reading this documentation will not know that `create()` has a chain of promises inside it. Use the terms which the audience know.

File: src/editor/editor.js [hunk 3]
```diff
@@ -104,6 +104,19 @@ export default class Editor {
...
+		 * * ready - after initializing chain, the editor is ready to work
```
Comment (Reinmar) [inline src/editor/editor.js]: Relate to things which the user knows. When is the editor ready? When the promise returned by `create()` is resolved.

File: src/editor/editor.js [hunk 4]
```diff
@@ -104,6 +104,19 @@ export default class Editor {
...
+		 * * destroyed - after destroy, editor and all plugins are destroyed
```
Comment (Reinmar) [inline src/editor/editor.js]: after the `editor.destroy()` method was called

**Comment by pjasiun** [comment]
I think that specific editors should not need to set state. Base editor class should listen on `dataReady` and update it automatically. It could also change in the `destroy` event listener to make it consistent.

**Comment by oskarwrobel** [comment]
> I think that specific editors should not need to set state. Base editor class should listen on dataReady and update it automatically. It could also change in the destroy event listener to make it consistent.

I was thinking about it but I assumed that you are not gonna like it :D

**Comment by pjasiun** [comment]
I was also not sure if it should be `state` or `status` ([like in CKEditor 4](https://docs.ckeditor.com/ckeditor4/latest/api/CKEDITOR_editor.html#property-status)), but with @Reinmar we agreed that `state` is better.

File: src/editor/editor.js [hunk 5]
```diff
@@ -104,6 +104,19 @@ export default class Editor {
 		 */
 		this.t = this.locale.t;
 
+		/**
+		 * Indicates editor initialization status.
+		 * The following statuses are available:
```
Comment (Reinmar) [inline src/editor/editor.js]: Still bad. MIssing blank lines before and after this paragraph.

File: src/editor/editor.js [hunk 6]
```diff
@@ -104,6 +104,19 @@ export default class Editor {
...
+		 * * initializing - during the editor initialization
+		 * * ready - after the promise returned by the {@link module:core/editor/editor~Editor.create editor create()}
+		 * method is resolved
+		 * * destroyed - after the {@link #destroy editor destroy()} method was called
```
Comment (Reinmar) [inline src/editor/editor.js]:
```
destroyed - after the {@link #destroy `editor.destroy()`} method was called.
```

!!!

**Comment by Reinmar** [comment]
I'm going to fix the docs ;>

File: src/editor/editor.js (hunk 5 above)
Comment (Reinmar) [reply src/editor/editor.js]: BTW, states aren't available. You don't choose one. The editor may be in one of the following states.

Comment (Reinmar) [reply src/editor/editor.js]: Actually – "may be" is not completely correct to :D "The editor is..."

Comment (oskarwrobel) [reply src/editor/editor.js]: That part comes from CKE4 docs :D

**Comment by Reinmar** [comment]
Hehe :) I wanted to comment that this doesn't seem to be a good change if we need to make it in all repos. I can see you limited it to the `Editor` class, which is great.

**Pull request merged**